include/zephyr/drivers/battery.h

```diff
@@ -0,0 +1,470 @@
+
+/**
+ * Read time in minutes left when discharging.
+ *
+ * @param capacity	Destination for remaining time in minutes.
+ * @return non-zero if error.
+ */
+typedef int (*battery_time_to_empty)(const struct device *dev, int *minutes);
+
+/**
+ * Read run time in minutes left when discharging.
+ *
+ * @param capacity	Destination for remaining time in minutes.
+ * @return non-zero if error.
+ */
+typedef int (*battery_run_time_to_empty)(const struct device *dev, int *minutes);
+
+/**
+ * Read time in minutes left to full capacity when charging.
+ *
+ * @param capacity	Destination for remaining time in minutes.
+ * @return non-zero if error.
+ */
+typedef int (*battery_time_to_full)(const struct device *dev, int *minutes);
+
+/**
+ * Calculate battery time in minutes, under an assumed current.
+ *
+ * @param rate		Current to use for calculation, in mA.
+ *			If > 0, calculates charging time; if < 0, calculates
+ *			discharging time; 0 is invalid and sets minutes=0.
+ * @param minutes	Destination for calculated time in minutes.
+ * @return non-zero if error.
+ */
+typedef int (*battery_time_at_rate)(const struct device *dev, int rate, int *minutes);
+
+/**
+ * Read battery charge cycle count.
+ *
+ * @param count		Destination for count.
+ * @return non-zero if error.
+ */
+typedef int (*battery_cycle_count)(const struct device *dev, int *count);
+
+/**
+ * Read battery manufacture date.
+ *
+ * @param year		Destination for year
+ * @param month		Destination for month
+ * @param day		Destination for day
+ * @return non-zero if error.
+ */
+typedef int (*battery_manufacture_date)(const struct device *dev, int *year, int *month, int *day);
+
+/**
+ * Read battery serial number.
+ *
+ * @param serial	Destination for serial number.
+ * @return non-zero if error.
+ */
+typedef int (*battery_serial_number)(const struct device *dev, int *serial);
+
+/**
+ * Read manufacturer name.
+ *
+ * @param dest		Destination buffer.
+ * @param size		Length of destination buffer in chars.
+ * @return non-zero if error.
+ */
+typedef int (*battery_manufacturer_name)(const struct device *dev, char *dest, int size);
+
+/**
+ * Read manufacturer name.
+ *
+ * This can be overridden to return a chip or board custom string.
+ *
+ * @param dest		Destination buffer.
+ * @param size		Length of destination buffer in chars.
+ * @return non-zero if error.
+ */
+typedef int (*get_battery_manufacturer_name)(const struct device *dev, char *dest, int size);
+
+/**
+ * Read device name.
+ *
+ * @param dest		Destination buffer.
+ * @param size		Length of destination buffer in chars.
+ * @return non-zero if error.
+ */
+typedef int (*battery_device_name)(const struct device *dev, char *dest, int size);
+
+/**
+ * Read battery type/chemistry.
+ *
+ * @param dest		Destination buffer.
+ * @param size		Length of destination buffer in chars.
+ * @return non-zero if error.
+ */
+typedef int (*battery_device_chemistry)(const struct device *dev, char *dest, int size);
```

Comment:
Can interfaces from line 237-291 just be a get_battery_info and return a single structure that has date, number, names and chemistry (maybe other things as well) in a single call?

Reply:
Sounds like a solid idea, bonus if done with an ifdef around everything involving that since a lot of folks using Zephyr won't really care or have that from a simple fuel gauge.

Reply:
I do think having a minimal api + a maximal one, controlled by CONFIG_, would be sensible. There is a lot in this api that could be helpful but also a lot of code/data that might not ever be wanted.
If this were done, consider if it be acceptable to have a code submission that only implemented minimal... is that ok, or not? (just asking, not telling).